007 crud races

jpisarek opened the pull request.

jpisarek committed:
Added endpoint races/create

jpisarek committed:
Fixed status code - added unauthorized

jpisarek committed:
Added endpoint GET: races (GetAll)

jpisarek committed:
Added endpoint GET: races/{raceId}

jpisarek committed:
Added endpoint PUT: races/{raceId}

jpisarek committed:
Fixed status codes

jpisarek committed:
Added endpoint DELETE: races/{raceId}

kobrynsky commented on iPetCare/API/Controllers/RaceController.cs:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        [Authorize(Roles = Role.Administrator)]
+        [HttpPost("")]
```
W teorii jak jest pusty to nie potrzebujemy ("")  😃

kobrynsky commented on iPetCare/API/Controllers/RaceController.cs:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        [AuthorizeRoles(Role.Administrator, Role.Vet, Role.Owner)]
+        [HttpGet("")]
```
Tak samo tutaj 😃

kobrynsky commented on iPetCare/Application/Dtos/Races/CreateDtoRequest.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Application.Dtos.Races
+{
+    public class CreateDtoRequest
```
Wprowadziliśmy nową konwencję dto:
{Controller}{NazwaAkcji}{Dto}[Request/Response}
Czyli tutaj: RaceCreateDtoRequest

kobrynsky commented on iPetCare/Application/Dtos/Races/CreateDtoRequest.cs:
```diff
@@ -0,0 +1,13 @@
+        public string Name { get; set; }
+
+        [Required]
+        public int SpeciesId { get; set; }
```
tutaj powinno być SpecieId (w liczbie pojedynczej 😄 )

JJMPdsk replied:
@blc132 https://brians.wsu.edu/2016/05/31/specie/

kobrynsky replied:
@JJMPdsk  
![image](https://user-images.githubusercontent.com/26911602/77162953-67b4a580-6aad-11ea-8d29-1c0d6a4766d4.png)

kobrynsky commented on iPetCare/Application/Dtos/Races/CreateDtoResponse.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace Application.Dtos.Races
+{
+    public class CreateDtoResponse
```
konwencja

kobrynsky commented on iPetCare/Application/Dtos/Races/CreateDtoResponse.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        public string Name { get; set; }
+        public int Id { get; set; }
+        public int SpeciesId { get; set; }
```
liczba pojedyncza

kobrynsky commented on iPetCare/Application/Dtos/Races/DeleteDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Application.Dtos.Races
+{
+    public class DeleteDtoResponse
```
konwencja

kobrynsky commented on iPetCare/Application/Dtos/Races/DeleteDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public string Name { get; set; }
+        public int Id { get; set; }
+        public int SpeciesId { get; set; }
```
liczba pojedyncza

kobrynsky commented on iPetCare/Application/Dtos/Races/GetAllDtoResponse.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Application.Dtos.Races
+{
+    public class GetAllDtoResponse
```
konwencja

kobrynsky commented on iPetCare/Application/Dtos/Races/GetAllDtoResponse.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public List<RaceGetAllDtoResponse> Races { get; set; }
+    }
+    public class RaceGetAllDtoResponse
```
konwencja

kobrynsky commented on iPetCare/Application/Dtos/Races/GetAllDtoResponse.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public string Name { get; set; }
+        public int Id { get; set; }
+        public int SpeciesId { get; set; }
```
liczba pojedyncza

kobrynsky commented on iPetCare/Application/Dtos/Races/GetDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Application.Dtos.Races
+{
+    public class GetDtoResponse
```
konwencja

kobrynsky commented on iPetCare/Application/Dtos/Races/GetDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public string Name { get; set; }
+        public int Id { get; set; }
+        public int SpeciesId { get; set; }
```
tutaj chcielibyśmy dostać jeszcze nazwę gatunku, bo samo Id za dużo nam nie mówi.
Dodałbym nowy dto "RaceGetSpecieDtoResponse" w którym znajdowałoby się Id i Name

JJMPdsk replied:
@blc132 lub po prostu dać tutaj dodatkowy SpeciesName i zmapować w mapperze. Dzięki temu mamy "jedno-poziomowy" DTO. Pytanie co lepsze

kobrynsky commented on iPetCare/Application/Dtos/Races/PutDtoRequest.cs:
```diff
@@ -0,0 +1,12 @@
+    public class PutDtoRequest
+    {
+        public string Name { get; set; }
+        public int SpeciesId { get; set; }
```
liczba pojedyncza

kobrynsky commented on iPetCare/Application/Dtos/Races/PutDtoRequest.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Application.Dtos.Races
+{
+    public class PutDtoRequest
```
konwencja nazewnictwa

kobrynsky commented on iPetCare/Application/Dtos/Races/PutDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Application.Dtos.Races
+{
+    public class PutDtoResponse
```
konwencja nazewnictwa

kobrynsky commented on iPetCare/Application/Dtos/Races/PutDtoResponse.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public string Name { get; set; }
+        public int Id { get; set; }
+        public int SpeciesId { get; set; }
```
liczba pojedyncza

kobrynsky commented on iPetCare/Application/Interfaces/IRaceService.cs:
```diff
@@ -0,0 +1,15 @@
+        Task<ServiceResponse<GetAllDtoResponse>> GetAllAsync();
+        Task<ServiceResponse<GetDtoResponse>> GetAsync(int raceId);
+        Task<ServiceResponse<PutDtoResponse>> PutAsync(int raceId, PutDtoRequest dto);
+        Task<ServiceResponse<DeleteDtoResponse>> DeleteAsync(int raceId);
```
Tak właściwie czy jest potrzeba, żeby zwracać Dtosa w przypadku usuwania? W moim odczuciu to nie, wystarczy tylko status code (czyli zwracalibyśmy Task<ServiceResponse>

jpisarek replied:
Nad tym, czy przy usuwaniu wgl jakiś dto jest potrzebny też się zastanawiałam i najpierw myślałam, że nie, a potem stwierdziłam, że może fajnie jakby się wyświetliło użytkownikowi "Rasa xyz została usunięta"

JJMPdsk replied:
Myślę, że to bardziej już kwestia frontendu jak zareaguje na status code 200 :)

kobrynsky commented on iPetCare/Application/Services/RaceService.cs:
```diff
@@ -0,0 +1,178 @@
+            if(race == null)
+                return new ServiceResponse<PutDtoResponse>(HttpStatusCode.BadRequest, "Nie istnieje taka rasa w bazie danych");
+
+            race.Name = dto.Name;
```
Trzeba rozważyć przypadek kiedy dto.Name będzie miało nazwę która znajduje się w bazie (ale oczywiście pobrana wcześniej rasa może mieć taką nazwę)

kobrynsky commented on iPetCare/Application/Services/RaceService.cs:
```diff
@@ -0,0 +1,178 @@
+                return new ServiceResponse<PutDtoResponse>(HttpStatusCode.BadRequest, "Nie istnieje taka rasa w bazie danych");
+
+            race.Name = dto.Name;
+            race.SpeciesId = dto.SpeciesId;
```
A co jeśli w dto.Species będzie Id, które nie istnieje w bazie?
W moim odczuciu należy pobrać gatunek i sprawdzić czy jest nullem czy nie i wtedy dopiero dodać

kobrynsky commented on iPetCare/Application/Services/RaceService.cs:
```diff
@@ -0,0 +1,178 @@
+
+            if (result > 0)
+            {
+                var responseDto = new PutDtoResponse()
```
a tutaj możemy użyć automappera (_mapper)

JJMPdsk commented on iPetCare/API/Controllers/RaceController.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+            var response = await _raceService.CreateAsync(dto);
+
+            if (response.StatusCode == HttpStatusCode.OK)
```
Tak się zastanawiam czy jeśli taki sam ciąg sprawdzający czy httpstatuscode ok, unauthorized itd. stosujemy w wielu miejscach, może by to wyciągnąć w jakąś funkcję czy coś? @blc132

kobrynsky replied:
No wlasnie to wydaje się być sensowne, pytanie tylko jak by to miało wyglądać. Imho to detal i na razie bym nie zajmował sobie głowy tym

JJMPdsk commented on iPetCare/API/Controllers/RaceController.cs:
```diff
@@ -0,0 +1,98 @@
+
+        [Authorize(Roles = Role.Administrator)]
+        [HttpPut("{raceId}")]
+        public async Task<ActionResult<PutDtoResponse>> PutRace(int raceId, PutDtoRequest dto)
```
Imo dziwna nazwa na edycję/update, ale to tylko estetyka :D

JJMPdsk replied:
+ Nazwa przyjmowanych i zwracanych dtosków niewiele mówi. 
Chodzi o to, że w aplikacji będziemy mieli wiele putów, wiele creatów (w różnych kontrolerach) itd i jak je odróżnić między sobą?

JJMPdsk reviewed (commented):
Ogółem dobra robota

jpisarek committed:
Changed file names - new convention

jpisarek committed:
Added species to Dto in GET: races/{raceId}

jpisarek committed:
Fixed PUT: races/{raceId} endpoint - checking if exist dto.Name and dto.SpeciesId in database

jpisarek committed:
Added LazyLoading to DataContext. Fixed GET: races/{raceId}

Pawel-Kot commented on iPetCare/Application/Services/RaceService.cs:
```diff
@@ -0,0 +1,182 @@
+            if (currentUserName != null)
+            {
+                var currentUser = await _userManager.FindByNameAsync(currentUserName);
+                if (currentUser != null && currentUser.Role != Role.Administrator)
```
Czy tu nie trzeba dodać "|| currentUser == null"?

JJMPdsk replied:
W zasadzie to chyba faktycznie powinniśmy sprawdzić czy user jest nullem jak @Pawel-Kot  pisze. Tylko że wtedy już nie forbidden ale unauthorized. To sprawdzenie dodałbym do poniższego ifa z currentUserName == null (linijka 44)

jpisarek committed:
Added "currentUser == null" in create race endpoint

jpisarek committed:
Merge branch 'dev' into 007-crud-races
